KAFKA-13310 : KafkaConsumer cannot jump out of the poll method, and the…

RivenSun2 opened the pull request:
Title: KafkaConsumer cannot jump out of the poll method, and cpu and traffic on the broker side increase sharply
description:  The local test has been passed, the problem described by jira can be solved

JIRA link : https://issues.apache.org/jira/browse/KAFKA-13310

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 committed:
fix issue : KafkaConsumer cannot jump out of the poll method, and the cpu and traffic on the broker side increase sharply

JIRA link : https://issues.apache.org/jira/browse/KAFKA-13310

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 reviewed (commented):
After the problem is reproduced locally, the test case has been passed, and the problem can be fixed

RivenSun2 commented:
@showuon please help review this PR. Thank you.

RivenSun2 committed:
fix this issue for AUTO_PATTERN and AUTO_TOPICS in cleanUpConsumedOffsets method

JIRA link : https://issues.apache.org/jira/browse/KAFKA-13310

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 reviewed (commented):
clean unKnownTopicPartitions for AUTO_PATTERN and AUTO_TOPICS

RivenSun2 commented:
Thanks @guozhangwang

3schwartz commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1065,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+
+            //Because toGiveUpTopicPartitions may receive `UnknownTopicOrPartitionException` when submitting their offsets.
+            //We are prepared to abandon them. The worst effect is that these partitions may repeatedly consume some messages
+            log.warn("Synchronous auto-commit of offsets {} will be abandoned", toGiveUpTopicPartitions);
```
I believe the messages is in correct - it is the whole partitions for which any offsets are not committed.

`log.warn("Synchronous auto-commit of offsets for partitions {} will be abandoned", toGiveUpTopicPartitions);`

RivenSun2 replied:
Thank you for your reply.
Can you simulate the case of jira and debug test it locally?

The `allConsumedOffsets` variable is a collection of all topicPartitions with a valid position in the consumer's local area.
SourceCode as below:
`Map<TopicPartition, OffsetAndMetadata> allConsumedOffsets = subscriptions.allConsumed();`

`public synchronized Map<TopicPartition, OffsetAndMetadata> allConsumed() {
        Map<TopicPartition, OffsetAndMetadata> allConsumed = new HashMap<>();
        assignment.forEach((topicPartition, partitionState) -> {
            if (partitionState.hasValidPosition())
                allConsumed.put(topicPartition, new OffsetAndMetadata(partitionState.position.offset,
                        partitionState.position.offsetEpoch, ""));
        });
        return allConsumed;
    }`


In fact, what `toGiveUpTopicPartitions` is filtered here are all topicPartitions that are ready to commitOffsets but may trigger `UnknownTopicOrPartitionException`

3schwartz replied:
My comment was related to the text - it sounds like you are given up offsets but it is partitions - hence I would add the partition word in the sentence

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -670,10 +662,11 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
         // commit offsets prior to rebalance if auto-commit enabled
-        maybeAutoCommitOffsetsSync(time.timer(rebalanceConfig.rebalanceTimeoutMs));
+        //The timer whose commitOffset timed out is no longer time.timer(rebalanceConfig.rebalanceTimeoutMs), and is changed to the timer passed by the customer
+        maybeAutoCommitOffsetsSync(pollTimer);
```
Did we have any null handling for the `pollTimer`? Maybe we can pass the old `time.timer(rebalanceConfig.rebalanceTimeoutMs)` for null case here?

RivenSun2 replied:
Thank you for your reminder. I suggest that when `pollTimer` is null, it is safer to pass `time.timer(0L)` to `maybeAutoCommitOffsetsSync` instead of passing the old `time.timer(rebalanceConfig.rebalanceTimeoutMs)`

`time.timer(0L)` indicates that if `allConsumedOffsets` is not empty, submit the offset synchronously at least once

RivenSun2 replied:
Because even if we have the `cleanUpConsumedOffsets` method, we may encounter other non-`UnknownTopicOrPartitionException` **retryable** type exception  in the `commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets, Timer timer)` method.
 If we pass in time.timer(rebalanceConfig. rebalanceTimeoutMs), it may trigger the case in jira again.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -505,7 +497,7 @@ public boolean poll(Timer timer, boolean waitForJoinGroup) {
 
                 // if not wait for join group, we would just use a timer of 0
-                if (!ensureActiveGroup(waitForJoinGroup ? timer : time.timer(0L))) {
+                if (!ensureActiveGroup(waitForJoinGroup ? timer : time.timer(0L), timer)) {
```
I don't think we should also pass `timer` when not `waitForJoinGroup` here. I think when not `waitForJoinGroup`, we should not do `onJoinPrepare`, so we should pass `time.timer(0L)`, too. What do you think?

RivenSun2 replied:
The second `timer` variable here is a customer-defined timer, which is passed to the underlying `onJoinPrepare` method to replace the old timer
maybeAutoCommitOffsetsSync(**time.timer(rebalanceConfig.rebalanceTimeoutMs)**)

Expect to retry submitting offsets as much time as possible within the timeout period set by the customer

guozhangwang replied:
The main idea of passing a `timer(0)` here is that if ever the `poll` call triggers the rejoin group protocol, we do not block on any of that procedure, but in an after thought I agree it overlooks the needs to auto-commit offset during re-join which would definitely need more than `timer(0)`, but also should be less than `max.poll.interval`.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -66,15 +66,7 @@
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
Any reason you change to import all classes under `java.util`? I think we should import what we used in this class only.

RivenSun2 replied:
Sorry, this may be automatically modified by my code tool. I will restore it later.

showuon reviewed (commented):
@RivenSun2 , thanks for the PR. Left some comments. Thanks.

RivenSun2 commented:
Thanks for your review @showuon 
I will re-commit code later. Hope you can continue to give advice

RivenSun2 committed:
1. fix NPE issue
2. Optimize the import of package

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 commented:
@showuon 
The latest code changes have been committed, if you have time, please review it.
Thanks.

RivenSun2 commented:
@showuon  @guozhangwang 
Today I saw the official release of Kafka 3.0. First of all, congratulations to Kafka for opening a new chapter!
But I tested the 3.0 version locally, and the issue described by jira still exists.
Can you further promote the process of this PR?
Thanks  :）

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -670,10 +671,11 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
         log.debug("Executing onJoinPrepare with generation {} and memberId {}", generation, memberId);
         // commit offsets prior to rebalance if auto-commit enabled
-        maybeAutoCommitOffsetsSync(time.timer(rebalanceConfig.rebalanceTimeoutMs));
+        //The timer whose commitOffset timed out is no longer time.timer(rebalanceConfig.rebalanceTimeoutMs), and is changed to the timer passed by the customer
```
This comment is not necessary IMO.

RivenSun2 replied:
change codes already commit

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1074,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
         }
     }
 
+    private void cleanUpConsumedOffsets(Map<TopicPartition, OffsetAndMetadata> willCommitOffsets) {
```
nit: Do you think the variable name `partitionOffsetsToBeCommited` is better than `willCommitOffsets`?

RivenSun2 replied:
change codes already commit

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1074,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+        Set<String> validTopics = metadata.fetch().topics();
+        Set<TopicPartition> toGiveUpTopicPartitions = new HashSet<>();
+
+        Iterator<Map.Entry<TopicPartition, OffsetAndMetadata>> iterator = willCommitOffsets.entrySet().iterator();
```
We only need the entry key, so it could be changed to `willCommitOffsets.keySet().iterator();`

RivenSun2 replied:
change codes already commit

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1074,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+
+            if (!validTopics.contains(entry.getKey().topic())) {
+
+                toGiveUpTopicPartitions.add(entry.getKey());
```
I don't think we need extra `toGiveUpTopicPartitions` to store the partitions to be deleted. We can log the warning message in L1103 here directly

RivenSun2 replied:
I think that we store` toGiveUpTopicPartitions` variable can let customers know more clearly which topicPartitions will give up submitting offsets

showuon replied:
OK, make sense.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1054,6 +1056,9 @@ private void doAutoCommitOffsetsAsync() {
         if (autoCommitEnabled) {
             Map<TopicPartition, OffsetAndMetadata> allConsumedOffsets = subscriptions.allConsumed();
+
+            cleanUpConsumedOffsets(allConsumedOffsets);
```
Could we do this after we have `UnknownTopicOrPartitionException` happened? I think this issue is rarely happened, we can "lazily" clean it up. So, we can move this line into below `catch` block. (and need to add an `UnknownTopicOrPartitionException` case)

RivenSun2 replied:
change codes already commit

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -343,7 +344,7 @@ protected synchronized long timeToNextHeartbeat(long now) {
      */
     public void ensureActiveGroup() {
-        while (!ensureActiveGroup(time.timer(Long.MAX_VALUE))) {
+        while (!ensureActiveGroup(time.timer(Long.MAX_VALUE), null)) {
```
Actually, I don't think the `null` here make sense. Now, we take `null` as `timer(0)` (one try), right? But in this case, we put `Long.MAX_VALE` to wait for a long time, I think we should still use the original `rebalanceTimeout` as timer in this case.

showuon replied:
Could we pass in a timer here directly, instead of null?

RivenSun2 replied:
change codes already commit

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -420,7 +423,7 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 // exception, in which case upon retry we should not retry onJoinPrepare either.
                 needsJoinPrepare = false;
-                onJoinPrepare(generation.generationId, generation.memberId);
+                onJoinPrepare(generation.generationId, generation.memberId, pollTimer == null ? time.timer(0L) : pollTimer);
```
I'm thinking the `pollTimer` should not be null here, which makes sense more. We should pass a non-null timer into methods.

RivenSun2 replied:
change codes already commit

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1121,7 +1121,7 @@ public void testNormalJoinGroupFollower() {
         }, syncGroupResponse(assigned, Errors.NONE));
 
-        coordinator.joinGroupIfNeeded(time.timer(Long.MAX_VALUE));
+        coordinator.joinGroupIfNeeded(time.timer(Long.MAX_VALUE),null);
```
We should pass in a reasonable timer as the 2nd parameter (just like the 1st parameter). Same comments to below.

RivenSun2 replied:
change codes already commit

showuon reviewed (commented):
Thanks for the update. Left some more comments. And also, please add tests for your change. Thank you.

RivenSun2 committed:
commit changeCodes by showuon review

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 commented:
@showuon Thank you for your review, I agree with most of your suggestions.
I have submitted a new commit.

RivenSun2 committed:
fix package auto-import issue by IDEA Editor

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -188,8 +188,9 @@ public AbstractCoordinator(GroupRebalanceConfig rebalanceConfig,
+     * @param pollTimer A Timer constructed by the poll() timeout time set by the customer
      */
-    protected abstract void onJoinPrepare(int generation, String memberId);
+    protected abstract void onJoinPrepare(int generation, String memberId, final Timer pollTimer);
```
Since this timer is not always from `poll`, do you think we can rename it to `offsetCommitTimer` or something? And same comments apply to the java doc and other places.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -188,8 +188,9 @@ public AbstractCoordinator(GroupRebalanceConfig rebalanceConfig,
      * cleanup from the previous generation (such as committing offsets for the consumer)
      * @param generation The previous generation or -1 if there was none
      * @param memberId The identifier of this member in the previous group or "" if there was none
+     * @param pollTimer A Timer constructed by the poll() timeout time set by the customer
```
Maybe change to "The timer for committing offsets synchronously"?

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1016,6 +1018,9 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
             if (future.failed() && !future.isRetriable())
                 throw future.exception();
 
+            if(future.exception() instanceof UnknownTopicOrPartitionException)
+                cleanUpConsumedOffsets(offsets);
+
             timer.sleep(rebalanceConfig.retryBackoffMs);
```
It's not good to put the `cleanUpConsumedOffsets` before sleep since the topics could be changed during sleep. Maybe we can set a flag and clean up before next request sending?
something like this:
``` java
boolean shouldCleanUpConsumedOffsets = false;
do {
  // check here
  if (shouldCleanUpConsumedOffsets) {
    cleanUpConsumedOffsets(offsets);
    shouldCleanUpConsumedOffsets = false;
  }
  RequestFuture<Void> future = sendOffsetCommitRequest(offsets);
...
 
  if(future.exception() instanceof UnknownTopicOrPartitionException)
    shouldCleanUpConsumedOffsets = true;
} while (timer.notExpired());
```

What do you think?

RivenSun2 replied:
I agree, but the check logic should be placed outside the do-while layer

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1054,6 +1059,7 @@ private void doAutoCommitOffsetsAsync() {
     private void maybeAutoCommitOffsetsSync(Timer timer) {
         if (autoCommitEnabled) {
             Map<TopicPartition, OffsetAndMetadata> allConsumedOffsets = subscriptions.allConsumed();
+
```
nit: additional new line

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1075,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
     }
 
+    private void cleanUpConsumedOffsets(Map<TopicPartition, OffsetAndMetadata> partitionOffsetsToBeCommitted) {
+
```
nit: additional new line

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1075,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+        Iterator<TopicPartition> iterator = partitionOffsetsToBeCommitted.keySet().iterator();
+
+        while (iterator.hasNext()) {
+
```
nit: additional new line

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1075,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+        }
+
+        if (toGiveUpTopicPartitions.size() > 0) {
+
```
nit: additional new line

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1075,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+                toGiveUpTopicPartitions.add(topicPartition);
+                iterator.remove();
+            }
+
```
nit: additional new line

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1075,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+        if (toGiveUpTopicPartitions.size() > 0) {
+
+            //Because toGiveUpTopicPartitions may receive `UnknownTopicOrPartitionException` when submitting their offsets.
+            //We are prepared to abandon them. The worst effect is that these partitions may repeatedly consume some messages
```
Do you think this is better?
//We might get`UnknownTopicOrPartitionException` after submitting their offsets due to topics been deleted. We should update the offsets list here. The worst effect is that we may keep retrying to commit the offsets for the topics not existed any more, before timeout reached.

RivenSun2 replied:
I agree

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1069,6 +1075,37 @@ private void maybeAutoCommitOffsetsSync(Timer timer) {
+            //Because toGiveUpTopicPartitions may receive `UnknownTopicOrPartitionException` when submitting their offsets.
+            //We are prepared to abandon them. The worst effect is that these partitions may repeatedly consume some messages
+            log.warn("Synchronous auto-commit of offsets {} will be abandoned", toGiveUpTopicPartitions);
+
```
nit: additional new line

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java:
```diff
@@ -278,7 +278,7 @@ public void testTimeoutAndRetryJoinGroupIfNeeded() throws Exception {
         try {
             Timer firstAttemptTimer = mockTime.timer(REQUEST_TIMEOUT_MS);
-            Future<Boolean> firstAttempt = executor.submit(() -> coordinator.joinGroupIfNeeded(firstAttemptTimer));
+            Future<Boolean> firstAttempt = executor.submit(() -> coordinator.joinGroupIfNeeded(firstAttemptTimer, mockTime.timer(0)));
```
In my opinion, even though this is unit test, it might be safer to set the timer more than 0 because there could be possible that other retriable exception be thrown while commiting offsets. How about `mockTime.timer(10)`? (And same to other places in tests)

showuon reviewed (commented):
@RivenSun2 , thanks for updating the PR. Left some more comments. But still, we need some tests for these changes in the PR, especially the change in `commitOffsetsSync`. Please add them. Thank you.

RivenSun2 committed:
commit reviewed codeChange

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 commented:
@showuon  thanks for your review, i commit new change codes.
The test of code changes, I passed the end-to-end test, the consumer is running normally, and the bug has been resolved.
I think the unit test of the commitOffsetsSync method can reuse the method in `ConsumerCoordinatorTest`
     @Test
     public void testRetryCommitUnknownTopicOrPartition() {
     ...
    }
thanks.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -994,11 +996,16 @@ public boolean commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets,
         if (offsets.isEmpty())
             return true;
 
+        boolean shouldCleanUpConsumedOffsets = !checkConsumedOffsetsAreValid(offsets);
```
Instead of letting the callee `maybeAutoCommitOffsetsSync` silently dropping the given offsets that no longer exists, I think we should let the callee to simply report unknown topics out to the callers after timer elapsed and let callers to handle them. The main reason is that unknownTopicOrPartition may just temporary and hence silently dropping them at the callee can lead to confusing behaviors. 

Note there are several callers:

* When re-join group via `maybeAutoCommitOffsetsSync`: if `commitOffsetsSync` throws the unknown topic or partition exception after exhausting retries within the timer, the caller `maybeAutoCommitOffsetsSync` would log it upon capturing the exception and then wrap it as a `InterruptException` still. Note that the in the next poll call since `needsJoinPrepare` is false already we would not try to call `onJoinPrepare` again and if the topic is indeed deleted, the rebalance would remove the partitions from the subscription later. 
* When customer called via `Consumer#commitSync` directly: we can directly throw the timeout exception to the customer caller still. Note that since we would continuously log the error for `UnknownTopicOrPartition` error, user debugging it would still learn about the root cause why this commit call timed out.

RivenSun2 replied:
1. When re-join group via `maybeAutoCommitOffsetsSync`, the default timer is `time.timer(rebalanceTimeoutMs)`.
 Because our consumers' consumption logic is sometimes heavier, we refer to the configuration of Kafka stream https://kafka.apache.org/documentation/#upgrade_10201_notable
Set max.poll.interval.ms to Integer.MAX_VALUE
Therefore, in the maybeAutoCommitOffsetsSync method, there will be an endless loop of retry submitting the offset within the Integer.MAX_VALUE time.
The impacts are: (1)rebalance cannot be completed, group consumption progress is suspended; (2)offsets  commit with a high frequency, and CPU consumption on the broker side increases sharply.

2. For the call to `Consumer#commitSync`, if we can’t tolerate silently cleaning up non-existent offsets, we still have to make sure that the timer passed in by `commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets, Timer timer) commitOffsetsSync(Map<TopicPartition, OffsetAndMetadata> offsets, Timer timer)` method is a controllable value, `max.poll.interval.ms` allows users to set by themselves, However, it is applied here without declaration, which is obviously out of control.
Or we can put `checkConsumedOffsetsAreValid` & `cleanUpConsumedOffsets` in method of `maybeAutoCommitOffsetsSync (Timer timer)` before calling `commitOffsetsSync`

RivenSun2 replied:
"change its signature as ensureActiveGroup(timer, waitOnMetadata)"

I think if you pass `waitOnMetadata` to `ensureActiveGroup`, it will make things more complicated, because
`waitOnMetadata? timer: time.timer(0L)` will need be executed multiple times in `ensureCoordinatorReady` and `joinGroupIfNeeded`.

guozhangwang replied:
1) I think the max.poll.interval.ms is no longer set to `Integer.MAX_VALUE` in 2.3.0 (https://cwiki.apache.org/confluence/display/KAFKA/KIP-442%3A+Return+to+default+max+poll+interval+in+Streams).

2) For `Consumer#commitSync`, the `default.api.timeout.ms` is used if user do not specify the timeout, not the `max.poll.interval.ms`.

guozhangwang replied:
> To add, I think it is necessary to automatically clean up non-existent offsets in the `onJoinPrepare` method

I would not worry too much about `poll(long)` since 1) it is already deprecated since 2.0 and would be removed soon, and the main reason is indeed that it's, by design, going to block forever for metadata refresh / auto committing.

> I think if you pass waitOnMetadata to ensureActiveGroup, it will make things more complicated

I agree it would be more complicated at the moment, but once we remove the deprecated `poll(long)` API we would be able to simplify that logic since we would not need `waitOnMetadata` any more.

guozhangwang reviewed (commented):
Thanks for the PR @RivenSun2 and sorry for being so late on the review! I took a look at it and have two meta comments (had some more detailed comments inlined):

* I think it's a bit risky to let the `commitSync` callee to silently drop offsets that are considered not exist since the `unknownTopicOrPartition` would be a transient error. Instead we should always throw the error from the callee and the callers to decide what to do with it: 1) for auto-commit in `consumer.poll`, then we can just let the current call to throw interrupted exception and then the next call would not trigger `onJoinPrepare` anymore and would continue to the rebalance, which would naturally cleanup the assignment and hence the subscription offsets to commit.

* Instead of passing two timers in `ensureActiveGroup`, I think we can just change its signature as `ensureActiveGroup(timer, waitOnMetadata)` (and similarly for `joinGroupIfNeeded`), the timer would be used in the calls to commit offsets, and then we rely on the flag `waitOnMetadata` in `ensureCoordinatorReady` and inside `joinGroupIfNeeded` when polling for initialized join-group.

WDYT?

guozhangwang commented:
BTW I have not thoroughly gone through the added testing, I also agree with @showuon 's comments about the test coverage

RivenSun2 commented:
@guozhangwang  
To add, I think it is necessary to automatically clean up non-existent offsets in the `onJoinPrepare` method
We now agree that offsetCommitTimer can be the pollTimer passed by the customer. But if the customer calls `poll(final long timeoutMs)`, the situation will be worse, and `time.timer(Long.MAX_VALUE)` will replace pollTimer.

` if (includeMetadataInTimeout) {
                    // try to update assignment metadata BUT do not need to block on the timer for join group
                    updateAssignmentMetadataIfNeeded(timer, false);
                } else {
                    while (!updateAssignmentMetadataIfNeeded(time.timer(Long.MAX_VALUE), true)) {
                        log.warn("Still waiting for metadata");
                    }
                }`

guozhangwang commented:
@RivenSun2 Thanks for the thorough thoughts, I think the main concern of letting `onJoinPrepare` to clean up non-existent offsets are that, the `onJoinPrepare` itself would never know if the unknown topic/partitions are due to permanent deletions or just transient hiccups, and hence it's quite risky to cleanup the offsets.

The issue you brought up, could be resolved in a slightly way though: 1) first as you also mentioned, the `poll` call timer should be set to a reasonable value than `INF` by default, 2) I think the idea of introducing a separate config would as well be covered by the existing `DEFAULT_API_TIMEOUT_MS_CONFIG`, that is, if it is triggered from auto-commit, the timer should be set as the min(DEFAULT_API_TIMEOUT_MS_CONFIG, poll timer), if it is triggered from `consumer.commitSync` directly, then it should be set as the `DEFAULT_API_TIMEOUT_MS_CONFIG` or user specified values.

guozhangwang commented:
@RivenSun2 I talked to @hachikuji offline about the best options to fix it in near term, and we feel the async-commit approach may be more appropriate here. But you'd need to be careful about not just trying once and give up immediately to continue the rebalance. Here's the status quo:

* We would try commit upto the configured rebalance.timeout, and if we exhaust that timeout but still cannot succeed (like in this case, keep getting unknown topic partition error), we would just log it and continue the rebalance.
* Note that we have a flag `needsJoinPrepare` in AbstractCoordinator which is set before the `onJoinPrepare` call, which means that if the call itself throws out error, upon the next `poll` we would not try to trigger `onJoinPrepare` again.

So to make async-commit work, here's a rough sketch of what we'd need to do:

* We keep a reference of the last commit response future sent as part of the `onJoinPrepare`.
* In `maybeAutoCommitOffsetsSync`, as we would rename it to `maybeAutoCommitOffsetsAsync`, we check if the response future is `null` or not; if it is `null` we just send out the request and get hold on the `future`. And then we call the networkClient.poll once and see if the `future` is completed. If yes and there's no error, we return `true` from `maybeAutoCommitOffsetsAsync` indicating it has suceeded, otherwise we return `false`.
* When `maybeAutoCommitOffsetsAsync` returns false, the `onJoinPrepare` would return false immediately as well, and the caller would then reset the `needsJoinPrepare` flag so that next time it would still trigger `onJoinPrepare`. And then return to the `poll` call.

By doing that, the `poll` call would not block on commit, but would return immediately after just one trial of the commit request, and the user may potentially call `poll` multiple times in order to complete the commit as part of the `onJoinPrepare` to continue the rebalance, but it would help resolving the longer than `poll` timeout blocking issues. As for the backing off, let's delegate that to KIP-580.

WDYT?

guozhangwang commented:
Thanks everyone for the valuable discussions, this is a tricky issue indeed and worth some more clarifications, on the status quo. In general there are a lot of issues we've observed with coupling the rebalance protocol with the `poll` calls, and in the long run we are thinking to move the rebalance as well as fetching records completely into the background thread, and hence caller `poll` would only be getting the buffered records --- I will submit a JIRA ticket for this soon --- but as for now, what I'm thinking is a short-term fix, and hence the goal is to fix it with as less scope and complexity as possible.

The status quo is that, when we are about to enter the rebalance we try to commit sync with a `rebalance.timeout`. That means: 

1) the time blocked on commit maybe LARGER than the timeout passed into `poll`. This is the main issue we'd like to address.

2) if the commit exhausted the timeout but still cannot succeed, we would "give up" with a warning log and still continue the rebalance protocol with the risk that the new host of those partitions may fetch stale committed offsets (or even no offsets at all, if there's no committed offsets before the rebalance happened). This is not ideal, but okay according to the at least once semantics --- note, with exactly once semantics we do not rely on consumer to commit offsets so EOS is not impacted by this behavior. So any solution we may propose, as long as it would not make further regression, would be fine.

Now here are my thoughts:

* it's risky to just always upper bound the commit timeout with the `poll` parameter value, since in practice many users may call `poll(0)` or with a very small millisecond parameter. This would mean that, we may be much more aggressive on giving up and continue, i.e. making 2) above more exaggerate. Imagine if there's a leader migration happening just at the same time of the rebalance, we would very easily give up the commit even with just one round-trip failure.
* it's also risky to always upper bound the commit timeout with the configured rebalance regardless of the `poll` timeout, which we have been discussed many times above and we all know its impact now.

So what's left seems to be that, we still make a reasonably "best effort" to commit offsets before continue the rebalance protocol, and that "best effort" should be somewhat irrelevant to the poll timeout, but also not blocking more than the timeout value itself. Hence it occurs to me that, the best effort for now could be that we still do not block for longer than the poll timeout --- i.e. we may even return early --- but we do not give up the commit either, and hence here async commit where could be potentially completed across multiple `poll` calls would be reasonable.

RivenSun2 commented:
@guozhangwang  @showuon  and @hachikuji  Thank you for paying so much attention to this issue and giving a lot of valuable discussions.

> Hence it occurs to me that, the best effort for now could be that we still do not block for longer than the poll timeout --- i.e. we may even return early --- but we do not give up the commit either, and hence here async commit where could be potentially completed across multiple poll calls would be reasonable.

After thinking about it, I think this is indeed a good solution in the short term.

If the rebalance protocol can be separated from the poll thread later, it will make the poll method more pure (only poll message from cache or send fetch request). This is a very challenging task, and a new thread may be introduced. Or you can combine heartbeat threads to try to achieve separation.

Thank you for the long-term active discussion. I also hope that the Kafka project will develop better and better.

RivenSun2 committed:
commit reviewed codeChange

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 commented:
According to our previous discussion, I re-modified and submitted the code, local debugging, regular expression subscription consumption, deleting or creating a topic that conforms to the regular expression, the consumer runs normally, the rebalance can be completed normally, and the consumer can continue to consume messages ；
In subscribe(Collection<String> topics) mode, when you delete a topic in the list, everything is normal
Please help approve this PR, thank you



At the same time recently because of this jira solution https://issues.apache.org/jira/browse/KAFKA-13398,
I encountered an abnormal situation when upgrading my broker version, and this abnormal situation can be reproduced in version 3.0.0
I analyzed and positioned this, and finally found RC.
I created a new Jira
https://issues.apache.org/jira/browse/KAFKA-13422
Please help review and give some suggestions
thanks

RivenSun2 committed:
commit test code
add test Method "testForceMetadataDeleteForPatternSubscriptionDuringRebalance()"

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 committed:
fix merge conflicts

Author: RivenSun2 <riven.sun@zoom.us>

Reviewers: Luke Chen <showuon@gmail.com>

RivenSun2 commented:
Hi @showuon and @guozhangwang 
Any suggestions for my last reply?
Thanks

RivenSun2 committed:
Merge branch 'apache:trunk' into trunk

guozhangwang commented:
Restarted jenkins build.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -692,10 +693,18 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
+    protected boolean onJoinPrepare(int generation, String memberId) {
         log.debug("Executing onJoinPrepare with generation {} and memberId {}", generation, memberId);
-        // commit offsets prior to rebalance if auto-commit enabled
-        maybeAutoCommitOffsetsSync(time.timer(rebalanceConfig.rebalanceTimeoutMs));
```
I think upon close(), we can also use `maybeAutoCommitOffsetsAsync` and then we can remove the whole function fo `maybeAutoCommitOffsetsSync`.

RivenSun2 replied:
I agree with your suggestion.

guozhangwang replied:
Actually this is a bad suggestion.. see https://github.com/apache/kafka/pull/11898/files

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1073,6 +1086,33 @@ private void doAutoCommitOffsetsAsync() {
         });
     }
 
+    private boolean maybeAutoCommitOffsetsAsync() {
```
It's a bit awkward to modify `onJoinPrepareAsyncCommitFuture` inside the `maybeAutoCommitOffsetsAsync` function since the function name itself indicate a general purpose, but specifically for join-prepare --- though I understand today it is indeed only used for that caller.

How about letting the `maybeAutoCommitOffsetsAsync` to return the future instead of the boolean, and then let the caller a.k.a. the `onJoinPrepare` today to check if the future is completed or not.

RivenSun2 replied:
I agree with your suggestion.

showuon replied:
There's already another method called `maybeAutoCommitOffsetsAsync`, but with `void` return type. Are you sure the compiling succeeded?

RivenSun2 replied:
In fact, only onJoinPrepare called the maybeAutoCommitOffsetsAsync method before.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1054,11 +1067,11 @@ public void maybeAutoCommitOffsetsAsync(long now) {
         }
     }
 
-    private void doAutoCommitOffsetsAsync() {
```
nit: I know this is not introduced by this PR, but could we rename it to `autoCommitOffsetsAsync` instead?

RivenSun2 replied:
I agree with your suggestion.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1073,6 +1086,33 @@ private void doAutoCommitOffsetsAsync() {
+        if (autoCommitEnabled) {
+            invokeCompletedOffsetCommitCallbacks();
+
+            if (onJoinPrepareAsyncCommitFuture == null)
```
I cannot understand this logic clearly.. my original thought is that, we do not reference the `onJoinPrepareAsyncCommitFuture` here at all, just create a future and return to the caller.

RivenSun2 replied:
Because we use asynchronous commitOffsets, and at the same time, in order not to block the Consumer#poll method
We cannot call ConsumerNetworkClient#poll(future)

The onJoinPrepareAsyncCommitFuture variable is introduced to deal with the "onJoinPrepareAsyncCommitFuture.isDone() == false" situation.

Maybe my consideration is complicated, this situation can be retryed by Consumer#poll next time.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1073,6 +1086,33 @@ private void doAutoCommitOffsetsAsync() {
 
+    private boolean maybeAutoCommitOffsetsAsync() {
+        if (autoCommitEnabled) {
+            invokeCompletedOffsetCommitCallbacks();
```
Why invoke the callbacks twice instead of only once below?

RivenSun2 replied:
my fault, already fix it

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1073,6 +1086,33 @@ private void doAutoCommitOffsetsAsync() {
+                onJoinPrepareAsyncCommitFuture = null;
+                return true;
+            }
+            if (onJoinPrepareAsyncCommitFuture.failed() && !onJoinPrepareAsyncCommitFuture.isRetriable())
```
I think we can move this check to the caller (onJoinPrepare).

showuon replied:
+1

RivenSun2 replied:
I agree with your suggestion.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -420,9 +421,14 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 // exception, in which case upon retry we should not retry onJoinPrepare either.
                 needsJoinPrepare = false;
-                onJoinPrepare(generation.generationId, generation.memberId);
+                if (!onJoinPrepare(generation.generationId, generation.memberId))
```
This logic is a bit complicated now, thinking about it twice I feel it may read simpler as this:

```
if (needsJoinPrepare) {
    try {
        if (onJoinPrepare(generation.generationId, generation.memberId))
            needsJoinPrepare = false;
        else
            return false;
    } finally {
            // if onJoinPrepare throws an exception, it would be from the rebalance listener.
            // next time we would then not retry {@code onJoinPrepare} any more but proceed the join-group procedure.
            needsJoinPrepare = false;
    }
}
```

RivenSun2 replied:
If you do
This goal should not be achieved:
```
if (!onJoinPrepare(generation.generationId, generation.memberId))
     needsJoinPrepare = true;
```

showuon replied:
You're right. How about this:
```java

if (needsJoinPrepare) {
    try {
        if (onJoinPrepare(generation.generationId, generation.memberId))
            needsJoinPrepare = false;
        else
            return false;
    } catch (KafkaException e) {
            // if onJoinPrepare throws an exception, it would be from the rebalance listener.
            // next time we would then not retry {@code onJoinPrepare} any more but proceed the join-group procedure.
            needsJoinPrepare = false;
            throw e;
    }
}
```

RivenSun2 replied:
I think it can be simplified like this
 ```
           if (needsJoinPrepare) {
                // need to set the flag before calling onJoinPrepare since the user callback may throw
                // exception, in which case upon retry we should not retry onJoinPrepare either.
                needsJoinPrepare = false;
                if (!onJoinPrepare(generation.generationId, generation.memberId)) {
                    needsJoinPrepare = true;
                    //should not initiateJoinGroup if needsJoinPrepare still is true
                    return false;
                }
            }
```

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -403,7 +404,7 @@ private void closeHeartbeatThread() {
      * @param timer Timer bounding how long this method can block
      * @throws KafkaException if the callback throws exception
-     * @return true iff the operation succeeded
+     * @return true if the operation succeeded
```
why this change?

RivenSun2 replied:
my fault, I thought it was a misspelling of a word, I will restore it

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -692,10 +693,18 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
+            onJoinPrepareAsyncCommitSucceeded = maybeAutoCommitOffsetsAsync();
+        } catch (Exception e) {
+            onJoinPrepareAsyncCommitFuture = null;
+            onJoinPrepareAsyncCommitSucceeded = true;
```
why does this exception case seem as `CommitSucceeded = true`?

RivenSun2 replied:
If the exception is of non-retryable type, decide to abandon to asyncCommit offsets

showuon replied:
So, 3 comments here:
1. If we return `true`, we won't do `onJoinPrepare` on next poll, and the offset commit will be skipped. Is that expected? 
2. the returned `future` null means no commit operated (since no auto commit). Could you add a comment here to explain the case?
3. The variable name `onJoinPrepareAsyncCommitSucceeded` might be better to change to `onJoinPrepareAsyncCommitCompleted` ? WDYT?

RivenSun2 replied:
1. Yes.
If (!autoCommitEnabled || coordinatorUnknown()), future will be null, I think in this case, we can give up committing offsets for rebalance
2&3. sure

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -151,6 +151,11 @@
             put(topic2, 1);
         }
     });
+    private MetadataResponse deletedMetadataResponse = RequestTestUtils.metadataUpdateWith(1, new HashMap<String, Integer>() {
```
Since this variable only used in the new added test, could we put it into the test?

RivenSun2 replied:
Sure

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -1053,6 +1058,49 @@ public void testForceMetadataRefreshForPatternSubscriptionDuringRebalance() {
     }
 
+    @Test
+    public void testForceMetadataDeleteForPatternSubscriptionDuringRebalance() {
```
Nice test to simulate the scenario you faced.

RivenSun2 replied:
In fact, only onJoinPrepare called the maybeAutoCommitOffsetsAsync method before.
Follow guozhang's suggestion , I will delete the maybeAutoCommitOffsetsSync(timer) method,
close() also calls maybeAutoCommitOffsetsAsync

showuon reviewed (commented):
Thanks for the update. Sorry for late review. Left some comments.

RivenSun2 committed:
commit code changes for review

Author: RivenSun2 riven.sun@zoom.us

RivenSun2 commented:
@guozhangwang @showuon 
Thanks for your helpful review.

I just commit a new change
If you have time, please browse it.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -420,7 +421,11 @@ boolean joinGroupIfNeeded(final Timer timer) {
-                onJoinPrepare(generation.generationId, generation.memberId);
+                if (!onJoinPrepare(generation.generationId, generation.memberId)) {
+                    needsJoinPrepare = true;
+                    //should not initiateJoinGroup if needsJoinPrepare still is true
```
nit: // return false when onJoinPrepare is waiting for committing offset

RivenSun2 replied:
ok,  I will add a comment here

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -692,10 +692,22 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
+                // consistent with async auto-commit failures, we do not propagate the exception
+                log.warn("Asynchronous auto-commit offsets failed: {}", future.exception().getMessage());
+                onJoinPrepareAsyncCommitSucceeded = true;
+            }
```
I think we should also log the error in `failed && isRetriable()` case

RivenSun2 replied:
sure

showuon reviewed (commented):
@RivenSun2 , thanks for the update. Left some comments.

RivenSun2 committed:
commit code changes for showuon's review

RivenSun2 commented:
@showuon  Thanks again.
please browse it when available.

RivenSun2 commented:
Hi @guozhangwang @showuon 
can you help to promote this PR?
Thanks.

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -692,10 +692,25 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
+        boolean onJoinPrepareAsyncCommitCompleted = false;
+        // async commit offsets prior to rebalance if auto-commit enabled
+        // and if auto-commit disable or the coordinatorUnknown is true, the future will be null,
+        // the asynchronous commit operation will not do.
```
// and if auto-commit disable or the coordinatorUnknown is true, the future will be 
// the asynchronous commit operation will not do.

--> // null future means no offset commit request sent, so it is still considered completed

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -692,10 +692,25 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
+                onJoinPrepareAsyncCommitCompleted = true;
+            } else if (future.failed()) {
+                // consistent with async auto-commit failures, we do not propagate the exception
+                log.warn("Asynchronous auto-commit offsets failed: {}", future.exception().getMessage());
```
Sorry, I found we already log the error in `autoCommitOffsetsAsync`. We should remove the logging here. And make the if condition simpler as: 
```
if (future == null || future.succeeded() || (future.failed() && !future.isRetriable())) {
    onJoinPrepareAsyncCommitCompleted = true;
}

showuon replied:
Actually, after the concise if statement above, we can add comments like this:
```
// return true when
// 1. future is null, which means no commit request sent
// 2. offset commit completed
// 3. offset commit failed with non-retriable error
```

showuon commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1073,22 +1092,14 @@ private void doAutoCommitOffsetsAsync() {
+            invokeCompletedOffsetCommitCallbacks();
+            return future;
+        } else
+            return null;
```
nit: remove `else` here, return null at the end.

showuon reviewed (commented):
@RivenSun2 , Left some comments. Besides the comments, I think you should add more tests for the `onJoinPrepared` change, since you change some logics there, like you can test the return value of ``onJoinPrepared` in different cases. Thank you.

RivenSun2 commented:
@showuon 
Thanks for your suggestion. I will add more tests.

RivenSun2 committed:
commit code changes for showuon's review

RivenSun2 commented:
@showuon  please browse the latest commit when available.
Thanks.

RivenSun2 commented:
Hi @showuon @guozhangwang 

please help check this PR when available. Thank you.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -1073,22 +1084,14 @@ private void doAutoCommitOffsetsAsync() {
     }
 
-    private void maybeAutoCommitOffsetsSync(Timer timer) {
+    private RequestFuture<Void> maybeAutoCommitOffsetsAsync() {
```
I think these two overloaded `maybeAutoCommitOffsetsAsync` functions can be merged: compared with the other that returns `void`, this function did the following more:

1) `client.pollNoWakeup();` -> this has already been called inside the `commitOffsetsAsync` callee at https://github.com/apache/kafka/pull/11340/files#diff-0029e982555d1fae10943b862924da962ca8e247a3070cded92c5f5a5960244fR982, so this is not needed.

2) `invokeCompletedOffsetCommitCallbacks();` -> I think this is not required in the async call.

As a result, we can just leave one `maybeAutoCommitOffsetsAsync` function that returns the future.

RivenSun2 replied:
I agree with you,thanks

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -692,10 +692,17 @@ private void validateCooperativeAssignment(final Map<String, List<TopicPartition
+        // 1. future is null, which means no commit request sent, so it is still considered completed
+        // 2. offset commit completed
+        // 3. offset commit failed with non-retriable error
+        if (future == null || future.succeeded() || (future.failed() && !future.isRetriable()))
```
For case 3, we should log an ERROR entry indicating that we are going to proceed with commit failed still and it may result in duplicated data being consumed, and for such cases we should notify the user.

guozhangwang reviewed (commented):
@RivenSun2 Thanks for the updated PR and sorry for the late final reviews.

I have only two minor comments left and otherwise it LGTM.

showuon reviewed (commented):
@RivenSun2 , thanks for adding tests. Sorry that I totally forgot this PR. I checked again, and overall LGTM! The jenkins tests didn't run successfully last time. Let's see if the tests pass after you address guozhangwang 's comments. Thank you.

RivenSun2 committed:
commit code changes for guozhangwang's review

RivenSun2 commented:
@guozhangwang  @showuon 
Thanks for your review, I just submitted some changes.
please help check again when available. Thank you.

showuon commented:
Let's see the jenkins test results. Thanks.

showuon commented:
@RivenSun2 , it looks like you break some tests. Might need you to check them. From the logs in the last 2 runs: 
https://ci-builds.apache.org/job/Kafka/job/kafka-pr/view/change-requests/job/PR-11340/15/console
https://ci-builds.apache.org/job/Kafka/job/kafka-pr/view/change-requests/job/PR-11340/16/console

They all die in `KafkaConsumerTest`. I can also see `ConsumerBounceTest` also failed. Could you check and fix them? Thank you.

RivenSun2 committed:
fix issue for ConsumerBounceTest testClose case failed

RivenSun2 commented:
@showuon  @guozhangwang 

The reason why KafkaConsumerTest testCase failed seems to be "Timeout has been exceeded", which further led to the "Shutdown in progress" exception

Reasons for ConsumerBounceTest testCase failure:
Following guozhanwang's suggestion, I removed the maybeAutoCommitOffsetsSync method and changed the consumerCoordinator#close method to call maybeAutoCommitOffsetsAsync(), which caused the kafkaConsumer#close method to end earlier than testCase expected.
I re-modified the method of this testcase
please help check again when available. Thanks

showuon commented:
@RivenSun2 , sorry, I think there are other tests cause the `clients:unitTest` timeout. I have a quick look, but can't find which tests cause it stuck. Could you check it? 
Here's the test results: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/view/change-requests/job/PR-11340/17/

RivenSun2 commented:
Hi @showuon 
I checked out a new branch `trunk_for_uptodate` from my `trunk` branch, and pulled the latest code of `apache:trunk`

Using Gradle 7.2, Java 11 and Scala 2.13.6.
When `Task :clients:test` is executed locally, the testCases of the class I modified can pass.

But the final log output will be stuck after a certain testCase is executed, but we don't know which testCase it is stuck in. Did not observe any STARTED event of testCase in the last line of the log.
Can you execute Task :clients:test locally? Will the same phenomenon occur?

RivenSun2 commented:
@showuon 
Sorry, I analyzed it carefully, because in the `ConsumerCoordinator#close(timer)` method, we changed to call `maybeAutoCommitOffsetsAsync()`, so the following code block will be executed
```
while (pendingAsyncCommits.get()> 0 && timer.notExpired()) {
                 ensureCoordinatorReady(timer);
                 client.poll(timer);
                 invokeCompletedOffsetCommitCallbacks();
             }
```
But in some testCases of KafkaConsumerTest, Time calls the `new MockTime( )` method, which causes `autoTickMs=0`, which in turn causes `Timer#update()` to never update currentTimeMs, which leads to blocking in the `ConsumerNetworkClient#poll(future, timer)` method.

RivenSun2 committed:
fix issue for KafkaConsumerTest testCase failed

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -2230,7 +2230,7 @@ public void testInvalidGroupMetadata() throws InterruptedException {
         initMetadata(client, Collections.singletonMap(topic, 1));
         KafkaConsumer<String, String> consumer = newConsumer(time, client, subscription, metadata,
-                new RoundRobinAssignor(), true, groupInstanceId);
+                new RoundRobinAssignor(), false, groupInstanceId);
```
I understand the mockTime change below. But why should we change the `autoCommit` flag here?

RivenSun2 replied:
Because this testInvalidGroupMetadata deliberately calls `new MockTime()` to make `autoTickMs=0`, I cannot set autoTickMs to non-zero here.
`() -> consumer.poll(Duration.ofSeconds(5))` this line of code will block, so that `assertThrows(ConcurrentModificationException.class,consumer::groupMetadata)` can pass,
Then call `consumer.wakeup()` to make `() -> consumer.poll(Duration.ofSeconds(5))` throw `WakeupException` and exit blocking.
`assertTrue(service.awaitTermination(10, TimeUnit.SECONDS))` can also be passed

RivenSun2 replied:
Since the `ConsumerCoordinator#close` method will call `maybeAutoCommitOffsetsAsync()`, we need to add an additional OffsetCommit prepareResponse in `testCloseManualAssignment` case.
The code has been submitted, please help check it, thanks @showuon

showuon replied:
I'm afraid the change of `autoCommit` flag will change the test purpose. I checked and it looks like we'll wait for consumer to close forever, because the timer didn't tick. Could we do `consumer.close(Duration.ZERO);`, and keep the `autoCommit` as true as before? What do you think?

showuon commented:
@RivenSun2 , thanks for the fix. The test ran successfully now! From the test results, there is a failed test: `testCloseManualAssignment`. Please fix it. Thanks.
REF: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11340/18/

RivenSun2 committed:
fix issue for testCloseManualAssignment case failed

RivenSun2 committed:
LeaveGroupResponse should be prepared conditionally

showuon commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -3202,17 +3202,18 @@ private void gracefulCloseTest(ConsumerCoordinator coordinator, boolean shouldLe
             OffsetCommitRequest commitRequest = (OffsetCommitRequest) body;
             return commitRequest.data().groupId().equals(groupId);
         }, new OffsetCommitResponse(new OffsetCommitResponseData()));
+        if (shouldLeaveGroup)
```
Could you explain why we need to check `shouldLeaveGroup` after your change?

RivenSun2 replied:
1.Sure, `consumer.close(Duration.ZERO)` is also a solution. Thanks.
My previous changes should not change the purpose of this test, and will not cause us to wait for the consumer to close forever. Because `autoCommit` is disabled, the `ConsumerCoordinator#close` method will end soon.
Local verification is also performed, `testInvalidGroupMetadata` case can be passed quickly and successfully

2.The `gracefulCloseTest` method is only called by two testCases, `testCloseDynamicAssignment` and `testCloseManualAssignment`.
`testCloseDynamicAssignment` uses the subscribe consumption model,
`testCloseManualAssignment` uses the assign consumption model. The difference here is: when `coordinator.close()` is executed, `testCloseDynamicAssignment` will initiate a LeaveGroupRequest in the underlying `AbstractCoordinator#maybeLeaveGroup(String leaveReason)` method, but `testCloseManualAssignment` will not.
So when I prepared `LeaveGroupResponse` here, I added a judgment on `shouldLeaveGroup`.
The previous `gracefulCloseTest` code logic should prepare redundant `LeaveGroupResponse` for `testCloseManualAssignment`.

showuon reviewed (commented):
@RivenSun2 , thanks for the update. Left some comments. Thanks.

RivenSun2 committed:
fix issue for testInvalidGroupMetadata case

showuon reviewed (approved):
@RivenSun2 , thanks for the update. The change LGTM now. But could you rebase to the latest `trunk` branch to trigger the jenkins build again, and see if the failed tests are failed due to our change. Thank you.

failed tests:
```
    Build / JDK 17 and Scala 2.13 / kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsByDurationToEarliest()
    Build / JDK 17 and Scala 2.13 / kafka.api.TransactionsBounceTest.testWithGroupId()
    Build / JDK 17 and Scala 2.13 / kafka.server.ReplicaManagerTest.[1] usesTopicIds=true
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.StreamsUncaughtExceptionHandlerIntegrationTest.shouldShutdownMultipleThreadApplication
    Build / JDK 8 and Scala 2.12 / kafka.server.KRaftClusterTest.testCreateClusterAndPerformReassignment()
    Build / JDK 11 and Scala 2.13 / kafka.admin.LeaderElectionCommandTest.[1] Type=Raft, Name=testPathToJsonFile, Security=PLAINTEXT
    Build / JDK 11 and Scala 2.13 / kafka.api.ConsumerBounceTest.testClose()
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.IQv2StoreIntegrationTest.verifyStore[cache=true, log=false, supplier=ROCKS_SESSION]
```

guozhangwang commented:
Retriggering jenkins tests.

RivenSun2 committed:
Merge branch 'apache:trunk' into trunk

showuon commented:
The failed tests are flaky and also failed sometimes in recent trunk jenkins build. Looks good.
```
    Build / ARM / kafka.server.ReplicaManagerTest.[1] usesTopicIds=true
    Build / JDK 17 and Scala 2.13 / integration.kafka.server.FetchRequestBetweenDifferentIbpTest.testControllerOldIBP()
    Build / JDK 17 and Scala 2.13 / kafka.metrics.MetricsTest.[2] quorum=kraft
```

RivenSun2 commented:
@guozhangwang 
Hi, guozhang, please help to check this PR.
Thanks a lot.

guozhangwang commented:
@RivenSun2 Thanks for the final commits and thanks to @showuon for making more passes as well. I just did another pass and it looks good to me too! I tried to resolve the conflicts myself and merge but seems it's a bit complicated. @RivenSun2 could you rebase again (sorry!!) and then I will merge.

RivenSun2 committed:
Merge branch 'trunk' of https://github.com/apache/kafka into trunk

 Conflicts:
	clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

RivenSun2 committed:
branch conflicts resolved

RivenSun2 commented:
Hi @guozhangwang 
I pulled the latest code of apache:trunk and resolved the conflict.
Please help to check again,thanks.

showuon commented:
@RivenSun2 , I checked the build results after your rebase, there's a test `ConsumerBounceTest.testClose` that failed twice with NPE. That looks suspicious and didn't fail in `trunk` branch.  Could you please take a look? Thanks.
```
[Build / JDK 8 and Scala 2.12 / kafka.api.ConsumerBounceTest.testClose()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11340/25/testReport/junit/kafka.api/ConsumerBounceTest/Build___JDK_8_and_Scala_2_12___testClose__/)
    [Build / JDK 8 and Scala 2.12 / kafka.api.ConsumerBounceTest.testClose()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11340/25/testReport/junit/kafka.api/ConsumerBounceTest/Build___JDK_8_and_Scala_2_12___testClose___2/)
```

guozhangwang commented:
> @RivenSun2 , I checked the build results after your rebase, there's a test ConsumerBounceTest.testClose that failed twice with NPE. That looks suspicious and didn't fail in trunk branch. Could you please take a look? Thanks.

+1. It seems relevant to the recent commit?

RivenSun2 committed:
in order to retrigger jenkins tests

RivenSun2 commented:
@guozhangwang  @showuon 
In fact, I ran ConsumerBounceTest.testClose() several times locally yesterday, and the testCase can all pass. I guess the jenkins environment is unstable, which occasionally leads to the return result of `consumer.committed(Set(tp).asJava).get(tp)`  is **null**.
I retriggered the jenkins tests

RivenSun2 commented:
`DynamicBrokerReconfigurationTest.testThreadPoolResize()`  and `RebalanceSourceConnectorsIntegrationTest.testRemovingWorker()`
The failed testCases are not related to this PR, and the failed testCases can pass locally.

guozhangwang commented:
Yeah I think I agree with you now, the failed cases for 2.12 is `connect.integration` and for 2.13 is `DynamicBrokerReconfiguration` which is known to be a bit flaky.

guozhangwang commented:
Merged to trunk, thanks @RivenSun2 !
